code/datums/extensions/lockable.dm

```diff
@@ -0,0 +1,216 @@
+			spark_system.set_up(5, 0, A.loc)
+			spark_system.start()
+			playsound(A.loc, 'sound/weapons/blade1.ogg', 50, 1)
+			playsound(A.loc, "sparks", 50, 1)
+			return TRUE
+
+		if(isScrewdriver(W))
+			if (do_after(user, 20 * user.skill_delay_mult(SKILL_DEVICES), holder))
+				user.show_message(SPAN_NOTICE("You [open ? "open" : "close"] the service panel."))
+				if(open)
+					close()
+				else
+					open()				
+			return TRUE
+
+		if(isMultitool(W) && open && !l_hacking)
+			user.show_message(SPAN_NOTICE("Now attempting to reset internal memory, please hold."), 1)
+			l_hacking = 1
+			if (do_after(user, 100 * user.skill_delay_mult(SKILL_ELECTRICAL), holder))
+				if (prob(user.skill_fail_chance(SKILL_DEVICES, 40, SKILL_EXPERT)))
+					l_setshort = FALSE
+					user.show_message(SPAN_NOTICE("Internal memory reset. Please give it a few seconds to reinitialize."), 1)
+					addtimer(CALLBACK(src, /datum/extension/lockable/proc/reset_memory), 3 SECONDS)
+					return TRUE
+				else
+					user.show_message(SPAN_WARNING("Unable to reset internal memory."), 1)
+					l_hacking = FALSE
+					bad_access_attempt(user)
+					return TRUE
+			else	
+				l_hacking = FALSE
+	return FALSE
+
+/datum/extension/lockable/proc/reset_memory()
+	if(!l_hacking)
+		return
+	reset_memory()
+	l_setshort = FALSE
+	l_hacking = FALSE
+	l_code = null
+	l_set = FALSE
+	var/atom/movable/A = holder
+	A.update_icon()
+
+/datum/extension/lockable/proc/emag_act(var/remaining_charges, var/mob/user, var/feedback)
+	var/atom/movable/A = holder
+	if(!emagged)
+		emagged = TRUE
+		locked = FALSE
+		to_chat(user, (feedback || "You short out the lock of \the [A]."))
+		A.update_icon()
+		return TRUE
+
+/datum/extension/lockable/storage
+	base_type = /datum/extension/lockable
+	expected_type = /obj/item/storage
+	
+
+/datum/extension/lockable/storage/open()
+	open = TRUE
+
+/datum/extension/lockable/storage/close()
+	open = FALSE
+
+/datum/extension/lockable/storage/safe
+	base_type = /datum/extension/lockable
+	expected_type = /obj/item/storage
+
+/datum/extension/lockable/charge_card	
+	base_type = /datum/extension/lockable
+	expected_type = /obj/item/charge_card
+	var/shock_strength = 0.2
+	var/alarm_loudness = 1
+
+/datum/extension/lockable/charge_card/New(holder, var/is_digital = FALSE)
+	..(holder, TRUE)
+	
+	var/obj/item/charge_card/E = holder
+	if(!istype(E))
+		return
+	switch(E.grade)
+		if("silver")
+			shock_strength = 0.4
+			max_code_length = 7
+			alarm_loudness = 3
+		if("gold")
+			shock_strength = 0.6
+			max_code_length = 9
+			alarm_loudness = 5
+		if("platinum")
+			shock_strength = 0.9
+			max_code_length = 11
+			alarm_loudness = 7
+
+/datum/extension/lockable/charge_card/bad_access_attempt(var/mob/user)
+	shock(user, 80)
+	var/atom/A = holder
+	A.audible_message(SPAN_WARNING("\The [holder] shrills in an annoying tone, alerting those nearby of unauthorized tampering."), hearing_distance = alarm_loudness)
```

Comment:
unless you add an actual sound, it'll be only heard by those in view, so pretty useless for anti-tampering.